cmd/skopeo/copy.go

```diff
@@ -99,23 +102,40 @@ See skopeo(1) section "IMAGE NAMES" for the expected format
 }
 
 // parseMultiArch parses the list processing selection
-// It returns the copy.ImageListSelection to use with image.Copy option
-func parseMultiArch(multiArch string) (copy.ImageListSelection, error) {
-	switch multiArch {
-	case "system":
-		return copy.CopySystemImage, nil
-	case "all":
-		return copy.CopyAllImages, nil
-	// There is no CopyNoImages value in copy.ImageListSelection, but because we
-	// don't provide an option to select a set of images to copy, we can use
-	// CopySpecificImages.
-	case "index-only":
-		return copy.CopySpecificImages, nil
-	// We don't expose CopySpecificImages other than index-only above, because
-	// we currently don't provide an option to choose the images to copy. That
-	// could be added in the future.
+// It returns the copy.ImageListSelection, instance list, and platform list to use in image.Copy option
+func parseMultiArch(multiArch string) (copy.ImageListSelection, []digest.Digest, []imgspecv1.Platform, error) {
+	switch {
+	case multiArch == "system":
+		return copy.CopySystemImage, nil, nil, nil
+	case multiArch == "all":
+		return copy.CopyAllImages, nil, nil, nil
+	case multiArch == "index-only":
+		// There is no CopyNoImages value in copy.ImageListSelection per se, but
+		// we can get the desired effect by using CopySpecificImages.
+		return copy.CopySpecificImages, nil, nil, nil
+	case strings.HasPrefix(multiArch, "sparse:"):
+		sparseArg := strings.TrimPrefix(multiArch, "sparse:")
+		platformSpecs := strings.Split(sparseArg, ",")
+		var platformList []imgspecv1.Platform
+		var instanceList []digest.Digest
+		for _, platformSpec := range platformSpecs {
+			if instanceDigest, err := digest.Parse(platformSpec); err == nil {
+				instanceList = append(instanceList, instanceDigest)
+				continue
+			}
+			if platformSpec == "system" {
+				platformList = append(platformList, imgspecv1.Platform{})
```

Comment:
`imgspecv1.Platform{}` does not generally mean “the current platform”.

… it does seem to mean that in the current implementation. To me, that just seems surprising and risky long-term.

(Compare https://github.com/containers/image/pull/1938 — I think _ideally_ I’d like the “turn `system` into a specific choice” logic to live in c/image.)

---

Actually, I don’t think this works as is. `skopeo copy --override-arch amd64 --multi-arch=sparse:system` should, I think, copy the amd64 image, not the image for `runtime.GOARCH` as the two PRs currently do if I’m reading them right. I’ll need to revisit this tomorrow…

Reply:
You're right - we choose the current platform for "system", ignoring `--override-arch`, which doesn't feel right.  Using a Platform built using overrideOS/overrideArch/overrideVariant seems to match the intent better.